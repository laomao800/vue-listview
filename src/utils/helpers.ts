import { VNode } from 'vue'
import isPlainObject from 'lodash/isPlainObject'
import isEmpty from 'lodash/isEmpty'
import isFunction from 'lodash/isFunction'
import get from 'lodash/get'
import { FilterField } from '~/types'

/**
 * 判断 node 是否为 Vue 内部的 VNode 类型
 */
export function isVNode(node: any): node is VNode {
  return (
    node !== null &&
    typeof node === 'object' &&
    hasOwn(node, 'componentOptions')
  )
}

/**
 * 根据映射表设置的结构提取数据并返回一个新对象
 *
 * dataMapping({
 *   result: {
 *     results: [1, 2, 3]
 *   }
 * }, {
 *   items: 'result.results'
 * })
 *
 * -> { items: [1, 2, 3] }
 *
 * @param {Object} data 目标数据
 * @param {Object} dataMap 映射表
 */
export function dataMapping(
  data: Record<string, any> = {},
  dataMap: Record<string, any> = {}
) {
  const result: Record<string, any> = {}

  if (isPlainObject(dataMap)) {
    Object.keys(dataMap).forEach((key) => {
      try {
        const dataKey = key.toString()
        const dataValue = get(data, dataMap[key])
        result[dataKey] = dataValue
      } catch (e) {}
    })
  }

  return result
}

/**
 * 判断值是否为搜索栏内合法的值，通过验证的值才可继续作为参数随请求提交
 */
export function isValidFieldValue(val: any): val is FilterField {
  return !(
    val === null ||
    val === undefined ||
    val === '' ||
    ((Array.isArray(val) || isPlainObject(val)) && isEmpty(val))
  )
}

/**
 * 判断一个值是否有效的搜索栏 field 配置项
 */
export function isValidFieldConfig(field: any) {
  return (
    field &&
    (hasOwn(field, 'type') ||
      isFunction(field.render) ||
      isFunction(field) ||
      isVNode(field))
  )
}

export function nodeParents(node: Element, selector: string) {
  const allMatchs = Array.from(document.querySelectorAll(selector))
  if (allMatchs.length === 0) {
    return null
  }
  function find(curNode: Element): Element | null {
    const parentNode = curNode.parentNode as Element | null
    if (parentNode) {
      if (allMatchs.includes(parentNode)) {
        return parentNode
      } else if (parentNode.parentNode) {
        return find(parentNode)
      }
    }
    return null
  }
  return find(node)
}

export function isPromise(obj: any) {
  return (
    !!obj &&
    (typeof obj === 'object' || typeof obj === 'function') &&
    typeof obj.then === 'function'
  )
}

export function hasOwn(obj: Record<string, unknown>, key: string) {
  return Object.prototype.hasOwnProperty.call(obj, key)
}

export function isDef(val: any) {
  return val !== undefined
}

export function hasRenderFn<T>(item: any): item is T {
  return isPlainObject(item) && isFunction(item.render)
}

export function ensurePromise<T>(data: T) {
  return isPromise(data) ? data : Promise.resolve(data)
}

export function noop() {
  // noop
}

export function resolveOptions(
  optionsConfig: any,
  done: (options: any[]) => void
) {
  let optionsPromise = null
  if (optionsConfig) {
    if (Array.isArray(optionsConfig)) {
      optionsPromise = Promise.resolve(optionsConfig)
    } else if (isFunction(optionsConfig)) {
      optionsPromise = ensurePromise(optionsConfig(done))
    } else if (isPromise(optionsConfig)) {
      optionsPromise = optionsConfig
    }
  }
  return optionsPromise
}
