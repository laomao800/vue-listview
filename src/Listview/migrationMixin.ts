import Vue from 'vue'
import get from '@/utils/getValue'
import camelCase from 'lodash/camelCase'
import { warn } from '@/utils/debug'
import { hasOwn } from '@/utils'

const insteadWarn = (type: string, oldName: string, newName: string) =>
  warn(
    `[Migration][${type}]: '${oldName}' is removed, use '${newName}' instead.`
  )

export default Vue.extend({
  mounted() {
    // Props
    const props = get(this, '$vnode.data.attrs', {})
    const migrationProps: Record<string, string> = {
      tableProps: 'contentProps',
      tableEvents: 'contentEvents',
      tableSelection: 'selection',
    }
    Object.entries(props).forEach(([propName, propValue]) => {
      propName = camelCase(propName)
      if (migrationProps[propName]) {
        insteadWarn('Props', propName, migrationProps[propName])
      }
      if (propName === 'filterFields') {
        if (
          Array.isArray(propValue) &&
          propValue.some((field: any) => hasOwn(field, 'get'))
        ) {
          insteadWarn('Props', 'filterFields.get()', 'transformRequestData()')
        }
      }
      if (propName === 'contentMessage') {
        if (hasOwn(propValue as any, 'message')) {
          insteadWarn('Props', 'contentMessage.message', 'contentMessage.text')
        }
      }
    })

    // Slots
    const migrationSlots: Record<string, string> = {
      'prepend-filterbar-submit': 'prepend-submit',
      'append-filterbar-submit': 'append-submit',
      'prepend-filterbar-more': 'prepend-more',
      'append-filterbar-more': 'append-more',
    }
    Object.keys(this.$slots).forEach((name) => {
      if (migrationSlots[name]) {
        insteadWarn('Slots', name, migrationSlots[name])
      }
    })
  },
})
