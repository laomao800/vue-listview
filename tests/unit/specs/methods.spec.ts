import { mount } from '@vue/test-utils'
import { wait } from '../helpers'
import Listview from '@/index'

describe('Methods', () => {
  it('resetFilter()', async () => {
    const wrapper = mount(Listview, {
      propsData: {
        filterModel: {
          text: '123',
          multiple: [1, 2, 3],
        },
        filterFields: [
          { type: 'text', model: 'text' },
          { type: 'multiple', model: 'multiple' },
        ],
      },
    })
    const vm = wrapper.vm as any
    const store = wrapper.findComponent({ ref: 'storeProvider' })
    vm.resetFilter()
    await wait()
    expect(store.vm.$props.filterModel).toEqual({
      text: undefined,
      multiple: [],
    })
  })

  it('setContentMessage("text")', async () => {
    const wrapper = mount(Listview)
    const vm = wrapper.vm as any
    vm.setContentMessage('text')
    await wait()
    expect(wrapper.find('.lv__message').html()).toMatchSnapshot()
  })

  it('setContentMessage("text", "error")', async () => {
    const wrapper = mount(Listview)
    const vm = wrapper.vm as any
    vm.setContentMessage('text', 'error')
    await wait()
    expect(wrapper.find('.lv__message').html()).toMatchSnapshot()
  })

  it('setContentMessage("text", "error", true)', async () => {
    const wrapper = await mount(Listview, {
      propsData: {
        requestHandler: () => ({
          result: { items: [{}], total: 1 },
          is_success: true,
        }),
      },
    })
    const vm = wrapper.vm as any
    vm.setContentMessage('text')
    await wait()
    expect(wrapper.find('.lv__message').exists()).toBeFalsy()
    vm.setContentMessage('text', null, true)
    await wait()
    expect(wrapper.find('.lv__message').html()).toMatchSnapshot()
  })
})
