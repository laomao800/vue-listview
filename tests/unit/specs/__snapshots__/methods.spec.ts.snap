// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Methods setContentMessage("text") 1`] = `<span class="lv__message"><span class="lv__message-icon"><i class="el-icon-warning"></i></span><span class="lv__message-text">text</span></span>`;

exports[`Methods setContentMessage("text", "error") 1`] = `<span class="lv__message lv__message--error"><span class="lv__message-icon"><i class="el-icon-error"></i></span><span class="lv__message-text">text</span></span>`;

exports[`Methods setContentMessage("text", "error", true) 1`] = `<span class="lv__message"><span class="lv__message-icon"><i class="el-icon-warning"></i></span><span class="lv__message-text">text</span></span>`;
